docs/madr/decisions/046-grpc-probes.md

```diff
@@ -0,0 +1,239 @@
+
+[inject.14.golden.yaml](https://github.com/kumahq/kuma/blob/master/pkg/plugins/runtime/k8s/webhooks/injector/testdata/inject.14.golden.yaml)
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  annotations:
+    kuma.io/virtual-probes: enabled
+    kuma.io/virtual-probes-port: "19000"
+    ...
+  labels:
+    run: busybox
+  name: busybox
+spec:
+  containers:
+  - name: kuma-sidecar
+    image: kuma/kuma-sidecar:latest
+    ...
+  - image: busybox
+    livenessProbe:
+      httpGet:
+        path: /8080/metrics
+        port: 19000
+      initialDelaySeconds: 3
+      periodSeconds: 3
+    name: busybox
+    readinessProbe:
+      httpGet:
+        path: /3001/metrics
+        port: 19000
+      initialDelaySeconds: 3
+      periodSeconds: 3
+  initContainers:
+  - name: kuma-init
+    image: kuma/kuma-init:latest
+  ...
+```
+
+And in the converted `Dataplane` object, we can see the two probes are both handled by the same Virtual Probe listener on port 19000:
+
+```yaml
+type: Dataplane
+mesh: default
+metadata:
+  creationTimestamp: null
+spec:
+  probes:
+    endpoints:
+      - inboundPath: /metrics
+        inboundPort: 8080
+        path: /8080/metrics
+      - inboundPath: /metrics
+        inboundPort: 3001
+        path: /3001/metrics
+    port: 19000
+  ...
+```
+
+## Considered Options
+
+**Option 1:** Transform gRPC probes to HTTP probes, merging all the HTTP probes into the existing HTTP based Virtual Probes listener
+
+**Option 2:** Allocate a new and separated port for each of the gRPC probes, and route them back to the application probe handling port accordingly. 
+
+### Proposed option: option 2 - allocate a new and separated port for each of the gRPC probes, and route them back to the application probe handling port accordingly 
+
+The only information we can use to differentiate the gRPC probes is the port number, we actually can not use this value to differ from requests for multiple gRPC probes. While the port number is included in the underlying HTTP2 request as `Host` header, it changes to the same value if we want to use one single virtual probe listener to support forwarding traffic for all gRPC probes.  
+
+We need to allocate a new port for each of them. To allocate listener ports, we'll need a new Pod annotation to support user specifying the range of ports to be used. This annotation, with name `kuma.io/virtual-probes-port-range`, will replace the existing `kuma.io/virtual-probes-port` annotation. The annotation should be optional. When user does not specify the annotation, the default range of ports will be used.
+
+gRPC probe requests are actually HTTP2 requests, which are based on TCP. So with separated virtual probe listener created for each probe, we can use TCP proxies with mTLS disabled in these listeners to forward requests back to application ports. Thus, on the `Dataplane` there is no need to introduce extra fields capturing properties of user defined gRPC probes, except the port number.
```

Comment:
I have a concern here. 

Why do people even have gRPC probes instead of HTTP probes? My guess would be that they want to HC the same port as their grpc application port. If this is on separate port then why even bother with gRPC.

Assuming that probe and app is on the same port if we just expose the port without explicit restriction to this specific gRPC service, we basically open the whole app without mTLS.

I think rewriting probes and exposing them without mTLS only works if we are very careful of not exposing the rest of the app

Reply:
Good point. To prevent security issues, we should use HCM to handle probe requests and only forward the health check endpoint.

Reply:
Updated.